### PR TITLE
Various networking fixes

### LINKED ISSUE #204: Don't require a custom `.cargo/config` to build
The dependency issue with openssl (previously missing dependencies on `user32-sys` and `gdi32-sys`) has been fixed, and yet multirust still requires a custom cargo config to build.

For some reason, the dylib dependency from `gdi32-sys` does not get passed to the linker, despite it being present in the build-script output.

To reproduce this issue, install the openssl binaries exactly as the appveyor script does, and set these environment variables (assuming 64-bit):

```
DEP_OPENSSL_INCLUDE=C:\OpenSSL\openssl-1.0.2d-vs2015\include64
OPENSSL_INCLUDE_DIR=C:\OpenSSL\openssl-1.0.2d-vs2015\include64
OPENSSL_LIB_DIR=C:\OpenSSL\openssl-1.0.2d-vs2015\lib64
OPENSSL_LIBS=libeay32MT:ssleay32MT
```

Then do a clean cargo build. The expected result is a ton of warnings about missing pdb files, followed by the relevent error, a missing symbol from `gdi32.lib`.

### LINKED ISSUE #243: Build openssl on appveyor
The way we are acquiring openssl in our windows builds is sub-optimal. Can we build this ourselves like on linux?

cc @alexcrichton 

In the long term, switching from hyper to curl and rust-crypto (for sha256) is probably best because that gets is HTTP proxy support (what that entails I'm not sure), but we need to tighten this up.

### PR DESCRIPTION
**Pull request by brson**
* Use the platform-native TLS stacks with [native-tls](https://github.com/sfackler/rust-native-tls), removing the OpenSSL dep on Windows.
* Do TLS hostname verification, via native-tls
* Remove the direct OpenSSL dep by switching to rust-crypto for sha256
* Upgrade OpenSSL to the latest version on Unix

Fixes #243, fixes #204, #344

r? @alexcrichton cc @sfackler @seanmonstar

### CODE REVIEW

**Comment by brson** [comment]
It appears that rust-crypto has a native component. I'll find a sha256 implementation that doesn't.

File: src/rustup-utils/src/lib.rs [hunk 1]
```diff
@@ -1,13 +1,15 @@
 #![feature(core_intrinsics)] // For type_name().
 #![feature(fundamental)]
+#![feature(step_by)]
```
Comment (alexcrichton) [inline src/rustup-utils/src/lib.rs]: :( I thought we wanted to go the other way

**Comment by alexcrichton** [comment]
Nice! r+ from me basically, but looks like the openssl crate is still getting compiled on MinGW which is causing a build failure. Maybe something's bringing it in by accident?

File: src/rustup-utils/src/lib.rs (hunk 1 above)
Comment (brson) [inline src/rustup-utils/src/lib.rs]: Yeah, it just happened to be used by the sha code and didn't seem worth messing with, but it should be easy to remove. I'll look at it.

**Pull request merged**